tls/s2n_client_hello.c

```diff
@@ -299,11 +326,35 @@ int s2n_process_client_hello(struct s2n_connection *conn)
     return S2N_SUCCESS;
 }
 
-int s2n_client_hello_recv(struct s2n_connection *conn)
+static int s2n_client_hello_process_cb_response(struct s2n_connection *conn, int rc)
 {
-    /* Parse client hello */
-    POSIX_GUARD(s2n_parse_client_hello(conn));
+    if (rc >= 0) {
+        if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_BLOCKING) {
+            if (rc) {
+                conn->server_name_used = 1;
+            }    
+            return 0;
+        }
+        if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_NONBLOCKING) {
+            conn->client_hello.callback_async_await = 1;
+            POSIX_BAIL(S2N_ERR_ASYNC_BLOCKED);
+        } 
```

Comment:
Do we need a shortcut, where in non-blocking mode you could immediately resume the handshake from the callback?

Reply:
well, though not documented (and hence not a client contract yet)  but one could change the callback mode (nonblocking->blocking) during the callback execution itself. 
We first evaluate the CH callback mode after the callback has been invoked.

Reply:
update: added the same in docs, short circuit works (added a test as well)